spec.html

```diff
@@ -8871,18 +8871,22 @@ <h1>[[Construct]] ( _argumentsList_, _newTarget_ )</h1>
     </emu-clause>
 
     <emu-clause id="sec-createbuiltinfunction" aoid="CreateBuiltinFunction">
-      <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
-      <p>The abstract operation CreateBuiltinFunction takes arguments _steps_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
+      <h1>CreateBuiltinFunction ( _stepsOrClosure_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
+      <p>The abstract operation CreateBuiltinFunction takes arguments _stepsOrClosure_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
       <emu-alg>
-        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification.
+        1. Assert: _stepsOrClosure_ is either an Abstract Closure, a set of algorithm steps or other definition of a function's behaviour provided in this specification.
         1. If _realm_ is not present, set _realm_ to the current Realm Record.
         1. Assert: _realm_ is a Realm Record.
         1. If _prototype_ is not present, set _prototype_ to _realm_.[[Intrinsics]].[[%Function.prototype%]].
-        1. Let _func_ be a new built-in function object that when called performs the action described by _steps_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
+        1. Let _func_ be a new built-in function object that when called performs the action described by _stepsOrClosure_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
         1. Set _func_.[[Realm]] to _realm_.
         1. Set _func_.[[Prototype]] to _prototype_.
         1. Set _func_.[[Extensible]] to *true*.
         1. Set _func_.[[ScriptOrModule]] to *null*.
+        1. If _stepsOrClosure_ is an Abstract Closure, then
+          1. Let _len_ be the largest number of named parameters in the definition of the Abstract Closure _stepsOrClosure_. Optional parameters (which are indicated with brackets: `[` `]`) or rest parameters (which are shown using the form &laquo;...name&raquo;) are not included in the parameter count.
+          1. Perform ! SetFunctionLength(_func_, _len_).
+          1. Perform ! SetFunctionName(_func_, *""*).
```

Comment:
I appreciate that this makes the current in-prose requirement

> Functions that are identified as anonymous functions use the empty String as the value of the "name" property.

into an explicit algorithm step.

However, this does have the effect of defining a creation order for these two properties, where currently I believe it is undefined. That means that using this mechanism for any user-observable function is technically normative.

(On the plus side, most engines seem to agree with this ordering:
```js
(async function () {
  return await {
    then: v => v(v),
  };
})().then(v => {
  console.log(Reflect.ownKeys(v));
});
```
returns `["length", "name"]` on at least JSC, SM, and V8.)

Reply:
This order was made normative in <https://github.com/tc39/ecma262/pull/1490>.

Reply:
#1490 does not imply an order for these properties for built-in functions, as far as I can tell. Am I misreading it?

Reply:
I very much doubt if the *intention* of #1490 was to leave the ordering of those properties different between builtin and user functions, but it certainly may not apply to builtin functions as written.

Reply:
For non-builtin functions, the order has been "length" then "name" since ES6. ("length" is defined as part of the operation that creates the function, and "name" is defined shortly thereafter, if it all.) (PR #1490 was originally going to define "name" as part of the function-creating operation, but pulled back from that.)

For built-in functions, the creation of those properties isn't specified via algorithm, but FWIW, [ECMAScript Standard Built-in Objects](https://tc39.es/ecma262/#sec-ecmascript-standard-built-in-objects) has discussed them in the order "length" then "name" since ES6.

Reply:
This has now been normatively specified in <https://github.com/tc39/ecma262/pull/2116>.